Added Zooming to Mouse-Position

BestVanRome opened the pull request:
Hi @abreheret,
for my purpuses I added a zooming function to your project.

I used it to annotate large pictures, where a fine pixel-labelling is demanded.

Hope it can serve someone else in the future.

**Main Changes:**
_image_canvas.h (added lines: 75-78):_
	
```
QSize _act_im_size             ;

private slots:
    void adjustScrollBars();
```

_image_canvas.cpp(added lines: 104 - 141 & 187-191):_  
```
added function implementation and calling it in scaleChanged.
The calling position is important to do after the repainting.
Also added a message in the statusbar to show actual mouse position.
```

Questions welcome!

BestVanRome committed:
added zooming to mouse_pos; still comments of test present

BestVanRome committed:
Zooming: Removed Debug-Comments; added Message in StatusBar for Mouspos X/Y

daandres commented on src/image_canvas.cpp:
```diff
@@ -34,248 +34,292 @@ ImageCanvas::~ImageCanvas() {
+
+    if(_scroll_parent->verticalScrollBar())
+    {
+        float_t posHeightRel = float_t (mPos.y()/imSize.height());  //Relation Mauspos zu Höhe des Images
```
use `auto` instead of `float_t`

BestVanRome replied:
This is what causes AppVeyor to crash.
As this is a cast to auto, which is forbidden by the visual-studio-compiler:
https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-error-c3537?view=msvc-160

So I'm changing it back to float_t.

daandres commented on src/image_canvas.cpp:
```diff
@@ -34,248 +34,292 @@ ImageCanvas::~ImageCanvas() {
+
+    if(_scroll_parent->horizontalScrollBar())
+    {
+        float_t posWidthRel = float_t (mPos.x()/imSize.width()); //Relation Mauspos zu Breite des Images
```
use `auto` instead of `float_t`

daandres commented on src/image_canvas.cpp:
```diff
@@ -34,248 +34,292 @@ ImageCanvas::~ImageCanvas() {
+    {
+        float_t posHeightRel = float_t (mPos.y()/imSize.height());  //Relation Mauspos zu Höhe des Images
+
+        //Ziel scrollBar an Position setzen, wo Mauszeiger ist
```
`// set scrollbar to mouse position`

daandres commented on src/image_canvas.cpp:
```diff
@@ -34,248 +34,292 @@ ImageCanvas::~ImageCanvas() {
+        QScrollBar * verticalScroll = _scroll_parent->verticalScrollBar();
+        double vertScrollSpace = verticalScroll->maximum() - verticalScroll->minimum(); // allg
+        verticalScroll->setValue(vertScrollSpace*posHeightRel);
+            //mögliche Alternative: QWheelEvent::angleDelta().y() -> see example!!! : https://doc.qt.io/qt-5/qwheelevent.html#angleDelta
```
`// alternative: QWheelEvent::angleDelta().y() -> see example!!! : https://doc.qt.io/qt-5/qwheelevent.html#angleDelta`

daandres commented on src/image_canvas.cpp:
```diff
@@ -34,248 +34,292 @@ ImageCanvas::~ImageCanvas() {
+    {
+        float_t posWidthRel = float_t (mPos.x()/imSize.width()); //Relation Mauspos zu Breite des Images
+
+        //Ziel scrollBar an Position setzen, wo Mauszeiger ist
```
// set scrollbar to mouse position

daandres reviewed (changes requested):
great work! 

i get flickering when resizing images, but it works. 
see inline comments for requested changes.

BestVanRome commented:
Ok, so shall I do another commit and do another pull-request or what would be the appropriate next steps for me to finish the pull-request?

daandres commented:
I'm not code owner. 
but I would suggest to add a new commit in this pull request.

BestVanRome committed:
added suggested changes from pull request

daandres commented on src/image_canvas.cpp:
```diff
@@ -120,22 +120,22 @@ void ImageCanvas::adjustScrollBars()
     if(_scroll_parent->verticalScrollBar())
     {
-        float_t posHeightRel = float_t (mPos.y()/imSize.height());  //Relation Mauspos zu Höhe des Images
+        auto posHeightRel = auto (mPos.y()/imSize.height());  //Relation Mauspos to Height of Image
```
`//Relation Mauspos to Height of Image

auto posHeightRel = float_t (mPos.y()/imSize.height()); `

daandres commented on src/image_canvas.cpp:
```diff
@@ -120,22 +120,22 @@ void ImageCanvas::adjustScrollBars()
     {
-        float_t posWidthRel = float_t (mPos.x()/imSize.width()); //Relation Mauspos zu Breite des Images
+        auto posWidthRel = auto (mPos.x()/imSize.width()); ////Relation Mauspos to Width of Image
 
```
`// Relation Mauspos to Width of Image

auto posWidthRel = float_t (mPos.x()/imSize.width());`

BestVanRome replied:
Yeah sorry about that :facepalm: . I've been to fast on that one :D

abreheret commented:
I will take a closer look as soon as I have time. there are already some updates going on on the dev branch that i need to merge, that said your code does not compile under windows in current state (https://ci.appveyor.com/project/abreheret/pixelannotationtool/builds/38772849#L96).

BestVanRome committed:
resolved 'auto-bug'

BestVanRome commented:
I don't have a windows environment. 
I can build and run it smoothly on my Ubuntu 20.04.
What's the issue now? 
Can I look into appveyor on my own?

BestVanRome commented:
Hi @abreheret,
any news? 
Shall I update smth?

BestVanRome committed:
Update image_canvas.cpp

Changed the auto -> float_t again as it caused crash on visual-studio-compiler: C3537 (cast to auto is forbidden)

BestVanRome commented:
Ou dammit :-/. What's the matter 😆 now?!

BestVanRome committed:
Update image_canvas.cpp

changed to a static_cast

BestVanRome committed:
Update image_canvas.cpp

Correcting syntax with auto-keyword

BestVanRome committed:
initial added slic to gui (but yet in seperet window

BestVanRome committed:
SlicAlgorithm Mask and Labels kann be viewed in normal QtWidget now. Also when changin ImageTabs the View of SlicParams for this special Image in the Window gets updated!

BestVanRome committed:
Merge branch 'master' of https://github.com/BestVanRome/PixelAnnotationTool

arbasli2 commented:
Hi guys, any news regarding merging this feature to master?
